src/miner.cpp

```diff
@@ -865,44 +872,37 @@ void BlockAssembler::addPackageTxs(int &nPackagesSelected, int &nDescendantsUpda
-                    if (auto feeEntry = evmFeeMap.find(addrKey); feeEntry != evmFeeMap.end()) {
-                        // Key already exists. We check to see if we need to prioritize higher fee tx
-                        const auto& lastFee = feeEntry->second;
-                        if (txResult.prepay_fee > lastFee) {
+                    const auto evmKey = std::make_pair(txResult.sender, txResult.nonce);
+                    if (evmTXFees.count(evmKey)) {
+                        const auto& gasFees = evmTXFees.at(evmKey);
+                        if (txResult.prepay_fee > gasFees) {
                             // Higher paying fee. Remove all TXs from sender and add to collection to add them again in order.
-                            auto addrTxs = evmAddressTxsMap[addrKey.address];
-                            RemoveTxIters(txIters(addrTxs));
-                            // Remove all fee entries relating to the address
-                            for (auto it = evmFeeMap.begin(); it != evmFeeMap.end();) {
+                            RemoveEVMTransactions(txIters(evmTXs[txResult.sender]));
+                            for (auto it = evmTXFees.begin(); it != evmTXFees.end();) {
                                 const auto& [sender, nonce] = it->first;
-                                if (sender == addrKey.address) {
-                                    it = evmFeeMap.erase(it);
+                                if (sender == txResult.sender) {
+                                    it = evmTXFees.erase(it);
                                 } else {
                                     ++it;
                                 }
                             }
-                            // Buggy code to fix below:
-                            checkedTX.erase(addrTxs[addrKey.nonce]->GetTx().GetHash());
-                            addrTxs[addrKey.nonce] = sortedEntries[i];
-                            auto count{addrKey.nonce};
-                            for (const auto& [nonce, entry] : addrTxs) {
+                            checkedTX.erase(evmTXs[txResult.sender][txResult.nonce]->GetTx().GetHash());
+                            evmTXs[txResult.sender][txResult.nonce] = sortedEntries[i];
+                            auto count{txResult.nonce};
+                            for (const auto& [nonce, entry] : evmTXs[txResult.sender]) {
                                 inBlock.erase(entry);
                                 checkedTX.erase(entry->GetTx().GetHash());
                                 replaceByFee.emplace(count, entry);
                                 ++count;
                             }
-                            evmAddressTxsMap.erase(addrKey.address);
-                            evm_remove_txs_by_sender(evmQueueId, addrKey.address);
+                            evmTXs.erase(txResult.sender);
+                            evm_remove_txs_by_sender(evmQueueId, txResult.sender);
```

Comment:
When we call evm_remove_txs_by_sender, the current pipeline is such that we are actually removing all transactions from that specific sender in the txqueue. This actually means that if for example we send 5 evm txs from 0 to 5 nonces, and then an evm tx of nonce 3 with higher prepay gas is submitted which replaces the evm tx nonce 3 - we are wiping all txs from 0 to 5 here. This test isnt actually covered in feature_evm, since the nonce replacement is nicely replacing nonce 0 in the test case which misses this specific edge case.

Reply:
Resolving as the current pipeline is correct. We are wiping all of our indexes whenever we replace the evm tx (both ancestors and descendants), and adding them into the replacebyfee index.